lib/active_merchant/billing/gateways/ct_payment.rb

```diff
@@ -0,0 +1,297 @@
+        add_operator_id(post, options)
+        add_invoice(post,0, options)
+        add_payment(post, credit_card)
+        add_customer_data(post, options)
+
+        response = commit('verifyAccount', post)
+        send_ack(response.authorization)
+        response
+      end
+
+      def store(credit_card, options={})
+        requires!(options, :email)
+        post = {}
+        post[:LanguageCode] = 'E'
+        post[:Name] = credit_card.name.rjust(50, ' ')
+        post[:Email] = options[:email].rjust(240, ' ')
+        add_operator_id(post, options)
+        add_payment(post, credit_card)
+        add_customer_data(post, options)
+
+        response = commit('recur/AddUser', post)
+        send_ack(response.authorization)
+        response
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((&?auth-api-key=)[^&]*)i, '\1[FILTERED]').
+          gsub(%r((&?token:)[^&]*)i, '\1[FILTERED]').
+          gsub(%r((&?cardNumber:)[^&]*)i, '\1[FILTERED]')
+      end
+
+      private
+
+      def add_terminal_number(post, options)
+        post[:MerchantTerminalNumber] = options[:merchant_terminal_number] || ' ' * 5
+      end
+
+      def add_money(post, money)
+        post[:Amount] = money.to_s.rjust(11,'0')
+      end
+
+      def add_operator_id(post, options)
+        post[:OperatorID] = options[:operator_id] || '0' * 8
+      end
+
+      def add_customer_data(post, options)
+        post[:CustomerNumber] = options[:customer_number] || '0' * 8
+      end
+
+      def add_address(post, creditcard, options)
+        if address = options[:billing_address] || options[:address]
+          post[:CardHolderAddress] = ("#{address[:address1]}
+                                      #{address[:address2]}").rjust(20, ' ')
+          post[:CardHolderPostalCode] = address[:zip].gsub(/\s+/, '').rjust(9, ' ')
+        end
+      end
+
+      def add_invoice(post, money,  options)
+        post[:CurrencyCode] = options[:currency] || (currency(money) if money)
+        post[:InvoiceNumber] = options[:order_id].rjust(12,'0')
+        post[:InputType] = 'I'
+        post[:LanguageCode] = 'E'
+      end
+
+      def add_payment(post, payment)
+        if payment.is_a?(String)
+          post[:Token] = split_authorization(payment)[3].strip
+        else
+          post[:CardType] = CARD_BRAND[payment.brand] || ' '
+          post[:CardNumber] = payment.number #.rjust(40,' ')
+          post[:ExpirationDate] = expdate(payment)
+          post[:Cvv2Cvc2Number] = payment.verification_value
+        end
+      end
+
+      def send_ack(authorization)
+        post = {}
+        post[:TransactionNumber] = split_authorization(authorization)[0]
+        commit('ack', post)
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def split_authorization(authorization)
+        authorization.split(';')
+      end
+
+      def commit(action, parameters)
```

Comment:
As far as I can tell, we always, without exception, call `send_ack` after `commit`, and always in the same way:

```ruby
response = commit(some_action, params)
send_ack(response.authorization)
response
```

Given that, I think it'd be worth thinking about just folding it directly into `commit`. It'd be one thing if we sometimes didn't send the `ACK`, but it seems like we literally always do. (Alternatively, if we do call `commit` without `send_ack` sometimes, or if we want some flexibility, I still suspect it'd be a win to just add a `..., send_ack: true)` param to `commit`.)

Reply:
It can't be folded into the commit, it would end up in an endless loop since the send_ack is also a commit.

Reply:
It can still be folded; you'd just need to split out `commit` from `commit_raw` (or the other direction, `commit_with_ack` v. `commit`). Or just go the `send_ack:` param route, in which case `send_ack` would pass `false.`

Reply:
Gotcha. Updated!